search/query_language_whitebox_test.go

```diff
@@ -0,0 +1,132 @@
+	ctx := migration.NewMigrationContext(context.Background())
+	s.DBTestSuite.PopulateDBTestSuite(ctx)
+}
+
+func (s *queryLanguageWhiteboxTest) SetupTest() {
+	s.clean = cleaner.DeleteCreatedEntities(s.DB)
+	testIdentity, err := testsupport.CreateTestIdentity(s.DB, "jdoe", "test")
+	require.Nil(s.T(), err)
+	s.modifierID = testIdentity.ID
+}
+
+func (s *queryLanguageWhiteboxTest) TearDownTest() {
+	s.clean()
+}
+
+func (s *queryLanguageWhiteboxTest) TestMinimalANDOperation() {
+	input := `
+	{"AND": [{"space": "openshiftio"},
+                 {"status": "NEW"}
+	]}`
+	fm := map[string]interface{}{}
+
+	// Parsing/Unmarshalling JSON encoding/json
+	err := json.Unmarshal([]byte(input), &fm)
+
+	if err != nil {
+		panic(err)
+	}
+	q := &Query{}
+
+	parseMap(fm, q)
+
+	openshiftio := "openshiftio"
+	status := "NEW"
+	expected := &Query{Name: "AND", Value: nil, Negate: false, Children: &[]*Query{
+		&Query{Name: "space", Value: &openshiftio, Negate: false, Children: nil},
+		&Query{Name: "status", Value: &status, Negate: false, Children: nil}},
+	}
+	assert.Equal(s.T(), expected, q)
+}
+
+func (s *queryLanguageWhiteboxTest) TestMinimalORandANDOperation() {
+	input := `
+	{"OR": [{"AND": [{"space": "openshiftio"},
+                         {"area": "planner"}]},
+	        {"AND": [{"space": "rhel"}]}]}`
+	fm := map[string]interface{}{}
+
+	// Parsing/Unmarshalling JSON encoding/json
+	err := json.Unmarshal([]byte(input), &fm)
+
+	if err != nil {
+		panic(err)
+	}
+	q := &Query{}
+
+	parseMap(fm, q)
+
+	openshiftio := "openshiftio"
+	area := "planner"
+	rhel := "rhel"
+	expected := &Query{Name: "OR", Value: nil, Negate: false, Children: &[]*Query{
+		&Query{Name: "AND", Value: nil, Negate: false, Children: &[]*Query{
+			&Query{Name: "space", Value: &openshiftio, Children: nil},
+			&Query{Name: "area", Value: &area, Children: nil}}},
+		&Query{Name: "AND", Value: nil, Negate: false, Children: &[]*Query{
+			&Query{Name: "space", Value: &rhel, Children: nil}}},
+	}}
+	assert.Equal(s.T(), expected, q)
+}
+
+func (s *queryLanguageWhiteboxTest) TestMinimalORandANDandNegateOperation() {
+	input := `
+	{"OR": [{"AND": [{"space": "openshiftio"},
+                         {"area": "planner"}]},
+			 {"AND": [{"space": "rhel", "negate": true}]}]}`
+	fm := map[string]interface{}{}
+
+	// Parsing/Unmarshalling JSON encoding/json
+	err := json.Unmarshal([]byte(input), &fm)
+
+	if err != nil {
+		panic(err)
+	}
+	q := &Query{}
+
+	parseMap(fm, q)
+
+	openshiftio := "openshiftio"
+	area := "planner"
+	rhel := "rhel"
+	expected := &Query{Name: "OR", Value: nil, Negate: false, Children: &[]*Query{
+		&Query{Name: "AND", Value: nil, Negate: false, Children: &[]*Query{
+			&Query{Name: "space", Value: &openshiftio, Children: nil},
+			&Query{Name: "area", Value: &area, Children: nil}}},
+		&Query{Name: "AND", Value: nil, Negate: false, Children: &[]*Query{
+			&Query{Name: "space", Value: &rhel, Negate: true, Children: nil}}},
+	}}
+	assert.Equal(s.T(), expected, q)
+}
```

Comment:
This is still a WIP. I was working on converting the native Go structure into SQL based on Pranav's code: https://gist.github.com/pranavgore09/34fb2793affd09e30e0525928364b837